Added the ability to use the Google Closure Compiler

dmyroniuk opened the pull request:
BroadleafCommerce/QA#3232
Added the ability to use the Google Closure Compiler or the YUICompressor to minify the javascript resources.

dmyroniuk committed:
BroadleafCommerce/QA#3232 - Added the ability to use the Google Closure Compiler instead of the YUICompressor to minify the javascript resources.

dmyroniuk committed:
BroadleafCommerce/QA#3232 - Renamed the properties of the Google Closure compiler.

cja769 commented on common/src/main/java/org/broadleafcommerce/common/config/BroadleafCommonConfig.java:
```diff
@@ -48,4 +65,64 @@ public JpaVendorAdapter blJpaVendorAdapter() {
+
+        throw new IllegalStateException();
+    }
+
```
All of this should be in an `@Conditional` on a closure-compiler class existing since we're making it an optional dependency.

cja769 replied:
Sorry if the code snippet is vague. I'm referring to all of the Google Closure configuration

cja769 commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationServiceImpl.java:
```diff
@@ -64,20 +80,45 @@
+    protected CompilerOptions compilerOptions;
+
+    protected SystemPropertiesService systemPropertiesService;
+
```
This should be `protected Environment environment`. Our system properties are added to the Spring environment so for consistency we've been moving away from injecting the `SystemPropertiesService` in favor of using `Environment`

cja769 commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationServiceImpl.java:
```diff
@@ -64,20 +80,45 @@
     protected boolean disableOptimizations;
 
+    @Value("${minify.closure.compiler.enabled}")
+    protected boolean closureCompilerEnabled;
```
Having a boolean to tell a service to do one of two very different implementations is a dead giveaway that we need to have 2 implementations of `ResourceMinificationService`. One that implements how we do it for YUICompressor and one for Google Closure Compiler. We can either create the bean `blResourceMinificationService` in a configuration file based on if Google Closure Compiler is on the classpath or not or we can add that `@Conditional` directly on the class an let component scanning dictate which one is created. If there's common logic between the two minification service then an Abstract class should be created that has the common logic implemented

cja769 commented on common/src/main/java/org/broadleafcommerce/common/config/BroadleafCommonConfig.java:
```diff
@@ -48,4 +65,64 @@ public JpaVendorAdapter blJpaVendorAdapter() {
+    public Compiler blClosureCompiler() {
+        return new Compiler();
+    }
+
```
I'm not really sure why this is a prototype bean. It seems that we only need one implementation.

cja769 commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationService.java:
```diff
@@ -20,24 +20,43 @@
+ * "ECMASCRIPT_2017, ECMASCRIPT_NEXT, NO_TRANSPILE</li>
+ *
+ *  <li>minify.closure.compiler.warningLevel - Warnings level. Possible values: QUIET, DEFAULT, VERBOSE</li>
+ * </ul>
```
Since we're creating multiple implementations of this class it's probably better to put the docs that are relevant to that implementation there instead of in the interface. The interface should have minification library agnostic information about what this interface is used for and what the methods do.

cja769 commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationService.java:
```diff
@@ -56,9 +75,11 @@
      */
+    @Deprecated
     boolean getAllowSingleMinification();
 
```
We probably need to make sure that it's fine but I would rather have this method just moved to the YUI compressor implementation. I don't know why this was on the interface in the first place.

cja769 commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationServiceImpl.java:
```diff
@@ -64,20 +80,45 @@
+    protected Compiler getClosureCompiler() {
+        return null;
+    }
+
```
I'm guessing this is the reason on why the `Compiler` was a prototype bean. I'm not really sure what advantage making the `Compiler` bean a prototype is with using `@Lookup` instead of simply creating a singleton bean and injecting it is.

cja769 commented on pom.xml:
```diff
@@ -1591,7 +1592,12 @@
+            <dependency>
+                <groupId>com.google.javascript</groupId>
+                <artifactId>closure-compiler</artifactId>
+                <version>${closure.compiler.version}</version>
```
I think this dependency and/or the YUI compressor should be optional dependency. I'm leaning towards just the closure-compiler being optional. This will give us a way to use @Conditionals to determine if we're using the closure-compiler to minify JS.

cja769 reviewed (dismissed):
There's a couple open-ended questions. One is I don't know the intent of 5.3. I don't know if it's getting merged into 5.4 and live like the rest of Broadleaf or if this is completely separate from the mainline. The reason this is important is because I've heard of multiple issues with YUICompressor with other clients so I think changing this is good to do. Second, I know that O'Reilly replaced YUICompressor in their project and they're targeting Java 7 so it leads my to believe that there's a library out there that's Java 7 compatible which would allow us to put this in 5.2. Third, I don't know if we're planning on fully replacing YUICompressor or having it live side-by-side with a new library. If we're replacing it then we need to make sure that it's indicated that it's deprecated (not that many clients are even aware of what library we even use for minification). 

Either way the main thing that needs to happen is to have 2 separate implementations of the ResourceMinificationService where the bean is created based on the existence of the google closure dependency. I think it's best to make the google closure dependency optional since the YUICompressor dependency is still needed for CSS compression therefore it'll always be on the classpath anyways so technically a client changing to using the google compressor would be optional.

cja769 commented:
I contacted O'Reilly to ask which library they replaced YUICompressor with and they actually replaced it with a Javascript build tool and just turned off Broadleaf's minification process. Additionally I talked to Phillip Verhyden about this and he confirmed what I was thinking. All of the 5.x line should be Java 7 compatible so we can't add the Google library since it requires Java 8.

Commit:
BroadleafCommerce/QA#3232 Refactored compatibility for YUICompressor and Google Closure Compiler so that Google Closure Compiler is used if the dependency is on the classpath. Also modified various things for backwards compatibility purposes and did performance tweaks so that using Google Closure Compiller has similar performance to YUICompressor

phillipuniverse commented on common/src/main/java/org/broadleafcommerce/common/resource/service/GoogleClosureJavascriptMinificationServiceImpl.java:
```diff
@@ -0,0 +1,134 @@
+ * "ECMASCRIPT6_TYPED (experimental), ECMASCRIPT_2015, ECMASCRIPT_2016, "
+ * "ECMASCRIPT_2017, ECMASCRIPT_NEXT, NO_TRANSPILE</li>
+ *
+ *  <li>minify.closure.compiler.warningLevel - Warnings level. Possible values: QUIET, DEFAULT, VERBOSE</li>
```
If you are documenting this, you should also include the default values for all of these that you have set.

phillipuniverse commented on common/src/main/java/org/broadleafcommerce/common/resource/service/JavascriptMinificationService.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.Reader;
+import java.io.Writer;
+
+public interface JavascriptMinificationService {
```
Can you `@author` and give a little something about this in a Javadoc like a link to look at YUI or Google Closure or something? just looks weird with no javadoc on this

phillipuniverse commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationServiceImpl.java:
```diff
@@ -107,44 +100,46 @@ public Resource minify(Resource originalResource) {
-        return new GeneratedResource(minifiedBytes, filename);
     }
     
+    // TODO Change method signature to throw a ResourceMinificationException in 5.4
```
I don't think that we should change this to throw an exception ever; we should just straight up handle it  right?

cja769 replied:
We handle it in the call that calls this method by just returning the original file. Technically we could feed the BufferedReader into the BufferedWriter but that seems pretty round about when the calling method can just return the original file.

phillipuniverse commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationServiceImpl.java:
```diff
@@ -107,44 +100,46 @@ public Resource minify(Resource originalResource) {
+        } catch (ResourceMinificationException e) {
+            // We're throwing and catching ResourceMinificationException for backwards compatibility. 
+            // In 5.4 this method will throw a ResourceMinificationException
+            throw new RuntimeException(e);
```
Just throw `IOException(e)` instead, somebody else could check the cause

phillipuniverse commented on common/src/main/java/org/broadleafcommerce/common/resource/service/YUIJavscriptionMinificationServiceImpl.java:
```diff
@@ -0,0 +1,116 @@
+ **/
+@Service("blJavascriptMinificationService")
+@ConditionalOnMissingClass(value = "com.google.javascript.jscomp.Compiler")
+public class YUIJavscriptionMinificationServiceImpl implements JavascriptMinificationService {
```
This is a typo

phillipuniverse commented on common/src/main/java/org/broadleafcommerce/common/resource/service/ResourceMinificationServiceImpl.java:
```diff
@@ -107,44 +100,46 @@ public Resource minify(Resource originalResource) {
         String type = getFileType(originalResource, filename);
         if (type == null) {
-            LOG.info("Unsupported minification resource: " + filename);
+            LOG.trace("Unsupported minification resource: " + filename);
```
Why not continue to `LOG.info` here?

cja769 replied:
Info is fine. I didn't realize that was the original logging level. Since ReferenceSite has info logging on it seemed extra loud since this method gets called for image resources as well as JS, and CSS

Commit:
BroadleafCommerce/QA#3232 Fixed spelling mistakes, added docs, and extra polish

- Renamed the YUI js minification service since it was misspelled
- Added Javadocs
- Changed logging level when we can't compress the file given back to the original level
- Changed RuntimeException to IOException so that it makes a little bit more sense

phillipuniverse commented:
:shipit:
